staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -893,3 +907,186 @@ func (m *mockTxnStore) Transaction(txns ...Transaction) *TransactionError {
 	m.succeedCount = len(txns)
 	return nil
 }
+
+func TestReplaceEvents(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, atomic := range []bool{true, false} {
+		t.Run(fmt.Sprintf("Atomic=%v", atomic), func(t *testing.T) {
+			clientgofeaturetesting.SetFeatureDuringTest(t, clientgofeaturegate.AtomicFIFO, atomic)
+			source := fcache.NewFakeControllerSource()
+			store := NewStore(DeletionHandlingMetaNamespaceKeyFunc)
+			t.Cleanup(func() {
+				source.Shutdown()
+			})
+
+			recorder := &eventRecorder{}
+			c := newInformer(store, InformerOptions{
+				ListerWatcher: source,
+				ObjectType:    &v1.Pod{},
+				ResyncPeriod:  0,
+				Handler:       recorder,
+			})
+			go c.RunWithContext(ctx)
+			if !WaitForCacheSync(ctx.Done(), c.HasSynced) {
+				t.Fatal("Timed out waiting for cache sync")
+			}
+			testReplaceEvents(t, ctx, source, recorder, store)
+		})
+	}
+}
+
+func testReplaceEvents(t *testing.T, ctx context.Context, source *fcache.FakeControllerSource, m *eventRecorder, store Store) {
+	// Set up 3 pods initially
+	pKeep := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-keep", Namespace: "default"}}
+	pMod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-mod", Namespace: "default", Labels: map[string]string{"ver": "1"}}}
+	pDel := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-del", Namespace: "default"}}
+
+	source.Add(pKeep)
+	source.Add(pMod)
+	source.Add(pDel)
+
+	// Wait for all 3 events to appear
+	if err := wait.PollUntilContextTimeout(ctx, 10*time.Millisecond, 2*time.Second, true, func(_ context.Context) (bool, error) {
+		m.historyLock.Lock()
+		defer m.historyLock.Unlock()
+		return len(m.history) == 3, nil
+	}); err != nil {
+		t.Fatalf("Failed initial sync: %v", err)
+	}
+
+	// Clear history so we only compare what happens during the RESYNC/REPLACE
+	m.clearHistory()
+
+	// Silent Drift (Events missed by controller)
+
+	// Modify: Update pod
+	pModUpdated := pMod.DeepCopy()
+	pModUpdated.Labels["ver"] = "2"
+	source.ModifyDropWatch(pModUpdated)
+
+	// Remove pod-del
+	source.DeleteDropWatch(pDel)
+
+	// Create pod-add
+	pAdd := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-add", Namespace: "default"}}
+	source.AddDropWatch(pAdd)
+
+	// Force Relist (Trigger AtomicReplace/Replace)
+	// This simulates the reconnection where the controller discovers the silent changes
+	source.ResetWatch()
```

Comment:
I don't see any direct side effects of the dropping watch. Could we just call Replace directly?